doquic.go

```diff
@@ -0,0 +1,132 @@
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+//
+// DNS-over-QUIC implementation
+//
+
+// DNS over Dedicated QUIC Connections
+// RFC 9250
+// https://datatracker.ietf.org/doc/rfc9250/
+
+package dnscore
+
+import (
+	"context"
+	"crypto/tls"
+	"io"
+	"net"
+	"time"
+
+	"github.com/miekg/dns"
+	"github.com/quic-go/quic-go"
+)
+
+func (t *Transport) sendQueryQUIC(ctx context.Context, addr *ServerAddr,
+	query *dns.Msg) (stream quic.Stream, t0 time.Time, rawQuery []byte, err error) {
+
+	udpAddr, err := net.ResolveUDPAddr("udp", addr.Address)
+	if err != nil {
+		return
+	}
+
+	udpConn, err := net.ListenUDP("udp", &net.UDPAddr{IP: net.IPv4zero, Port: 0})
+	if err != nil {
+		return
+	}
+
+	tr := &quic.Transport{
+		Conn: udpConn,
+	}
+
+	// 1. Fill in a default TLS config and QUIC config
+	hostname, _, err := net.SplitHostPort(addr.Address)
+	if err != nil {
+		return
+	}
+	tlsConfig := &tls.Config{
+		NextProtos: []string{"doq"},
+		ServerName: hostname,
+	}
+	quicConfig := &quic.Config{}
+
+	// 2. Use the context deadline to limit the query lifetime
+	// as documented in the [*Transport.Query] function.
+	if deadline, ok := ctx.Deadline(); ok {
+		_ = udpConn.SetDeadline(deadline)
+	}
+
+	// RFC 9250
+	// 4.2.1.  DNS Message IDs
+	// When sending queries over a QUIC connection, the DNS Message ID MUST
+	// be set to 0.
+	query.Id = 0
+	rawQuery, err = query.Pack()
+	if err != nil {
+		return
+	}
+
+	t0 = t.maybeLogQuery(ctx, addr, rawQuery)
+
+	quicConn, err := tr.Dial(ctx, udpAddr, tlsConfig, quicConfig)
+	if err != nil {
+		return
+	}
+
+	stream, err = quicConn.OpenStream()
+	if err != nil {
+		return
+	}
+	stream.Write(rawQuery)
```

Comment:
I think what we are doing here is not fully compliant with [the RFC](https://datatracker.ietf.org/doc/rfc9250/) and, in particular, with this requirement laid out in Sect. 4.2.:

> In order for multiple responses to be parsed, a 2-octet length field is used in exactly the same way as the 2-octet length field defined for DNS over TCP [RFC1035].  The practical result of this is that the content of each QUIC stream is exactly the same as the content of a TCP connection that would manage exactly one query.
>
> All DNS messages (queries and responses) sent over DoQ connections MUST be encoded as a 2-octet length field followed by the message content as specified in [RFC1035].

We already have code to implement this kind of exchange with the server, in `dotcp.go` and we're using it to implement DNS over TCP and DNS over TLS. In fact, DNS over TLS is just a tiny function that setups the TLS connection and then invokes the `*Transport.queryStream` function.

In light of this, I was wondering whether it could be sensible to attempt to use the same code for DNS over QUIC as well. I think doing this would reduce the amount of duplicate code, but there will be some required changes (I am mostly guessing here, since I did not try this out systematically, just skimmed through the code):

1. We may want to change the `queryStream` function signature to take a specific interface (maybe named `dnsStream`?) rather than a `net.Conn`, and this interface should only specify the functions we *need* inside `queryStream` (and using a different interface would also communicate that `queryStream` accepts data types beyond the classical `net.Conn` and indeed could also accept a QUIC stream).

2. The `queryStream` function should have a way to *only* close the stream/conn when we're doing DoQ, which means DoQ requires us to complicate the original algorithm by a single `if`, which seems ~fine.

3. It seems to me `quic.Stream` does not have the `LocalAddr` and `RemoteAddr` method, so we should probably play composition by creating a new ad-hoc type that uses the `quic.Conn` or the `quic.Stream` to implement the required methods of `dnsStream` (or however we want to call this interface) to ensure the combination of `quic.Stream` and `quic.Conn` meets the requirements of the `queryStream` method.